### PR TITLE
[Keymap] Add ollyhayes personal keymap

### PR DESCRIPTION
**Pull request by ollyhayes**
## Description

Add my keymap for the crkbd keyboard

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/crkbd/keymaps/ollyhayes/README.md
Comment (drashna) [inline keyboards/crkbd/keymaps/ollyhayes/README.md:1]: please rename this to be lower case.

File: keyboards/crkbd/keymaps/ollyhayes/config.h
Comment (drashna) [inline keyboards/crkbd/keymaps/ollyhayes/config.h:21]:
```suggestion
```

File: keyboards/crkbd/keymaps/ollyhayes/config.h
Comment (drashna) [inline keyboards/crkbd/keymaps/ollyhayes/config.h:28]:
```suggestion
```

File: keyboards/crkbd/keymaps/ollyhayes/config.h
Comment (drashna) [inline keyboards/crkbd/keymaps/ollyhayes/config.h:32]: Could you explicitly define which animations are enabled? 
```suggestion
#define RGBLIGHT_EFFECT_BREATHING
#define RGBLIGHT_EFFECT_RAINBOW_MOOD
#define RGBLIGHT_EFFECT_RAINBOW_SWIRL
#define RGBLIGHT_EFFECT_SNAKE
#define RGBLIGHT_EFFECT_KNIGHT
#define RGBLIGHT_EFFECT_CHRISTMAS
#define RGBLIGHT_EFFECT_STATIC_GRADIENT
#define RGBLIGHT_EFFECT_RGB_TEST
#define RGBLIGHT_EFFECT_ALTERNATING
#define RGBLIGHT_EFFECT_TWINKLE
```

File: keyboards/crkbd/keymaps/ollyhayes/config.h
Comment (drashna) [inline keyboards/crkbd/keymaps/ollyhayes/config.h:46]:
```suggestion
#   define RGB_DISABLE_WHEN_USB_SUSPENDED // turn off effects when suspended
```

File: keyboards/crkbd/keymaps/ollyhayes/keymap.c
Comment (drashna) [inline keyboards/crkbd/keymaps/ollyhayes/keymap.c:21]: Please don't include c files.  You should use rules.mk to add it (`SRC += oled.c`).  Header files should be included, only.

File: keyboards/crkbd/keymaps/ollyhayes/oled.c
Comment (drashna) [inline keyboards/crkbd/keymaps/ollyhayes/oled.c:107]:
```suggestion
```

File: keyboards/crkbd/keymaps/ollyhayes/oled.c
Comment (drashna) [inline keyboards/crkbd/keymaps/ollyhayes/oled.c:110]:
```suggestion
    oled_write_ln(get_u8_str(rgb_matrix_get_mode(), ' '), false);
```

File: keyboards/crkbd/keymaps/ollyhayes/oled.c
Comment (drashna) [inline keyboards/crkbd/keymaps/ollyhayes/oled.c:83]:
```suggestion
```

File: keyboards/crkbd/keymaps/ollyhayes/oled.c
Comment (drashna) [inline keyboards/crkbd/keymaps/ollyhayes/oled.c:73]:
```suggestion
```

File: keyboards/crkbd/keymaps/ollyhayes/oled.c
Comment (drashna) [inline keyboards/crkbd/keymaps/ollyhayes/oled.c:86]:
```suggestion
    oled_write_ln(get_u16_str(ticks, ' '), false);
```

File: keyboards/crkbd/keymaps/ollyhayes/oled.c
Comment (drashna) [inline keyboards/crkbd/keymaps/ollyhayes/oled.c:95]:
```suggestion
```

File: keyboards/crkbd/keymaps/ollyhayes/oled.c
Comment (drashna) [inline keyboards/crkbd/keymaps/ollyhayes/oled.c:98]:
```suggestion
    oled_write_ln(get_u16_str(presses_str, ' '), false);
```

File: users/ollyhayes/ollyhayes.c
Comment (drashna) [inline users/ollyhayes/ollyhayes.c:19]:
```suggestion
```

File: users/ollyhayes/ollyhayes.c
Comment (drashna) [inline users/ollyhayes/ollyhayes.c:30]:
```suggestion
    tap_code16(C(S(KC_U)));
```

File: users/ollyhayes/ollyhayes.c
Comment (drashna) [inline users/ollyhayes/ollyhayes.c:36]:
```suggestion
    if (layer_state_cmp(state, BASE)) {
```

File: users/ollyhayes/ollyhayes.c
Comment (drashna) [inline users/ollyhayes/ollyhayes.c:38]:
```suggestion
    } else if (layer_state_cmp(state,GAMES)) {
```

File: users/ollyhayes/ollyhayes.c
Comment (drashna) [inline users/ollyhayes/ollyhayes.c:86]: `C(KC_LEFT)`

File: users/ollyhayes/ollyhayes.c
Comment (drashna) [inline users/ollyhayes/ollyhayes.c:95]: `C(KC_RIGHT)`

File: users/ollyhayes/ollyhayes.c
Comment (drashna) [inline users/ollyhayes/ollyhayes.c:24]: Why not just use unicode?  You're essentially duplicating the code here, anyways. 

https://docs.qmk.fm/#/feature_unicode

**Review by drashna — CHANGES_REQUESTED** [review]

File: users/ollyhayes/ollyhayes.c
Comment (drashna) [reply users/ollyhayes/ollyhayes.c:23]: Specifically, you can use `send_unicode_string(str)` for this instead.  You'd want to enable unicode, and you could use `UNICODE_COMMON = yes` in your rules, and just use the core unicode framework.  You'd just need to set the mode.

Comment (ollyhayes) [reply users/ollyhayes/ollyhayes.c:23]: Thanks! I switched to using the unicode feature and added the commit to the review.

**Comment by ollyhayes** [comment]
@drashna, I think I've addressed all your comments, and also added a few other things, and I rebased against the latest upstream/master. I've left them as separate commits so you can see what I did, shall I squash them all together?

I just have one other question, I never managed to get my 12 underglow leds to solder properly so I just left them out and shorted the data line, and then I changed `g_led_config`  in my branch to match the new set of leds like this: https://github.com/ollyhayes/qmk_firmware/commit/525cfdec3af10bd7ea793d65dcde8cb5afdc55dd

Obviously I can't submit this commit as it is, unless there an easy way to redefine this variable only for my keymap? If not I will just leave this commit only in my local repository.

**Review by drashna — APPROVED** [review]

File: keyboards/crkbd/keymaps/ollyhayes/keymap.c
Comment (fauxpark) [inline keyboards/crkbd/keymaps/ollyhayes/keymap.c:26]:
```suggestion
  [BASE] = LAYOUT_split_3x6_3(
// +------------+------+------+----------+----------+---------+--------+----------+----------+--------+---------+---------+
     NUM_SWITCH , KC_Q , KC_W , KC_F     , KC_P     , KC_G    , KC_J   , KC_L     , KC_U     , KC_Y   , KC_SCLN , KC_BSPC ,
     KC_ESC     , KC_A , KC_R , KC_S     , KC_T     , KC_D    , KC_H   , KC_N     , KC_E     , KC_I   , KC_O    , KC_ENT  ,
     KC_LSFT    , KC_Z , KC_X , KC_C     , KC_V     , KC_B    , KC_K   , KC_M     , KC_COMM  , KC_DOT , KC_SLSH , KC_RSFT ,
                                KC_LCTL , MO(PUNC) , MO(ARR) , KC_SPC , MO(PUNC) , KC_RCTL
```

File: keyboards/crkbd/keymaps/ollyhayes/keymap.c
Comment (fauxpark) [inline keyboards/crkbd/keymaps/ollyhayes/keymap.c:44]:
```suggestion
  [GAMES] = LAYOUT_split_3x6_3(
//  +---------+------+------+---------+---------+----------+---------+---------+---------+--------+---------+---------+
	  KC_TRNS , KC_Q , KC_W , KC_E    , KC_R    , KC_T     , KC_Y    , KC_U    , KC_I    , KC_O   , KC_P    , KC_TRNS ,
	  KC_TRNS , KC_A , KC_S , KC_D    , KC_F    , KC_G     , KC_H    , KC_J    , KC_K    , KC_L   , KC_SCLN , KC_TRNS ,
	  KC_TRNS , KC_Z , KC_X , KC_C    , KC_V    , KC_B     , KC_N    , KC_M    , KC_COMM , KC_DOT , KC_SLSH , KC_TRNS ,
							  KC_TRNS , KC_LALT , KC_SPACE , MO(ARR) , KC_TRNS , KC_TRNS
```

File: keyboards/crkbd/keymaps/ollyhayes/keymap.c
Comment (fauxpark) [inline keyboards/crkbd/keymaps/ollyhayes/keymap.c:62]:
```suggestion
  [FUNCTIONS] = LAYOUT_split_3x6_3(
//  +---------------+---------+---------+---------+---------+---------+---------+---------+---------+-------+---------+---------+
	  KC_TRNS       , KC_TRNS , KC_TRNS , KC_F8   , KC_F9   , KC_F12  , KC_F12  , KC_F7   , KC_F8   , KC_F9 , KC_TRNS , KC_TRNS ,
	  TG(FUNCTIONS) , KC_TRNS , KC_F4   , KC_F5   , KC_F6   , KC_F11  , KC_F11  , KC_F4   , KC_F5   , KC_F6 , KC_TRNS , KC_TRNS ,
	  KC_TRNS       , KC_TRNS , KC_F1   , KC_F2   , KC_F3   , KC_F10  , KC_F10  , KC_F1   , KC_F2   , KC_F3 , KC_TRNS , KC_TRNS ,
										  KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Comment by ollyhayes** [comment]
I've rebased onto the latest master and squashed it all into one commit, hope that's okay!

I think I've address all the comments, is there anything else I should do before this could be merged?

Thanks for all your help :slightly_smiling_face:

**Review by keyboard-magpie — APPROVED** [review]

**Review by fauxpark — APPROVED** [review]

**Pull request merged**
